src/parser_yin.c

```diff
@@ -0,0 +1,520 @@
+            LY_CHECK_RET(ret);
+            ret = lyxml_get_string(xml_ctx, data, &buf, &buf_len, &out, &out_len, &dynamic);
+            LY_CHECK_RET(ret);
+            *arg_val = lydict_insert(xml_ctx->ctx, out, out_len);
+            LY_CHECK_ERR_RET(!(*arg_val), LOGMEM(xml_ctx->ctx), LY_EMEM);
+        } else if (arg_type == YIN_ARG_NONE) {
+            continue;
+        } else {
+            /* unrecognized or unexpected attribute */
+            /* TODO add name of attribute to error message */
+            if (name) {
+                LOGERR(xml_ctx->ctx, LY_EDENIED, "Invalid attribute");
+                return LY_EVALID;
+            }
+        }
+    }
+
+    return LY_SUCCESS;
+}
+
+/**
+ * @brief Parse content of whole element as text.
+ *
+ * @param[in] xml_ctx Xml context.
+ * @param[in] data Data to read from.
+ * @param[out] value Where content of element should be stored.
+ */
+LY_ERR
+parse_text_element(struct lyxml_context *xml_ctx, const char **data, const char **value)
+{
+    LY_ERR ret = LY_SUCCESS;
+    char *buf = NULL, *out = NULL;
+    size_t buf_len = 0, out_len = 0;
+    int dynamic;
+
+    ret = yin_parse_attribute(xml_ctx, data, YIN_ARG_NONE, NULL);
+    LY_CHECK_RET(ret);
+    LY_CHECK_RET(xml_ctx->status != LYXML_ELEM_CONTENT, LY_EVALID);
+
+    if (xml_ctx->status == LYXML_ELEM_CONTENT) {
+        ret = lyxml_get_string(xml_ctx, data, &buf, &buf_len, &out, &out_len, &dynamic);
+        LY_CHECK_RET(ret);
+        *value = lydict_insert(xml_ctx->ctx, out, out_len);
+        LY_CHECK_ERR_RET(!(*value), LOGMEM(xml_ctx->ctx), LY_EMEM);
+    }
+
+    return LY_SUCCESS;
+}
+
+static LY_ERR
+yin_parse_revision_date(struct lyxml_context *xml_ctx, const char **data, char *rev, struct lysp_ext_instance **exts)
+{
+    LY_ERR ret = LY_SUCCESS;
+    const char *temp_rev;
+
+    if (rev[0]) {
+        LOGVAL_YANG(xml_ctx, LY_VCODE_DUPSTMT, "revision-date");
+        return LY_EVALID;
+    }
+
+    LY_CHECK_RET(yin_parse_attribute(xml_ctx, data, YIN_ARG_DATE, &temp_rev))
+    LY_CHECK_RET(ret != LY_SUCCESS, ret);
+    LY_CHECK_RET(lysp_check_date((struct lys_parser_ctx *)xml_ctx, temp_rev, strlen(temp_rev), "revision-date") != LY_SUCCESS, LY_EVALID);
+
+    strcpy(rev, temp_rev);
+    lydict_remove(xml_ctx->ctx, temp_rev);
+    /* TODO extension */
+
+    return ret;
+}
+
+LY_ERR
+yin_parse_import(struct lyxml_context *xml_ctx, const char *module_prefix, const char **data, struct lysp_import **imports)
+{
+    LY_ERR ret = LY_SUCCESS;
+    enum yang_keyword kw;
+    struct lysp_import *imp;
+    const char *prefix, *name;
+    size_t prefix_len, name_len;
+
+    /* allocate sized array for imports */
+    LY_ARRAY_NEW_RET(xml_ctx->ctx, *imports, imp, LY_EVALID);
+
+    /* parse import attributes  */
+    LY_CHECK_RET(yin_parse_attribute(xml_ctx, data, YIN_ARG_MODULE, &imp->name));
+
+    while ((ret = lyxml_get_element(xml_ctx, data, &prefix, &prefix_len, &name, &name_len) == LY_SUCCESS && name != NULL)) {
```

Comment:
What if the XML element contains some (unexpected/additional/non-YANG) attributes?